File: frontend/src/api/tickets.ts
import api from "./client";
import type { Ticket, TicketCreate, TicketList, TicketFilters, TicketStats, TicketUpdateStatus } from "../types/ticket";

export async function fetchStats(): Promise<TicketStats> {
  const { data } = await api.get<TicketStats>("/tickets/stats");
  return data;
}

export async function fetchTickets(filters: TicketFilters): Promise<TicketList> {
  const params: Record<string, string | number> = {};
  if (filters.status) params.status = filters.status;
  if (filters.priority) params.priority = filters.priority;
  if (filters.search) params.search = filters.search;
  if (filters.sort_by) params.sort_by = filters.sort_by;
  if (filters.order) params.order = filters.order;
  if (filters.page) params.page = filters.page;
  if (filters.limit) params.limit = filters.limit;

  const { data } = await api.get<TicketList>("/tickets", { params });
  return data;
}

export async function createTicket(ticket: TicketCreate): Promise<Ticket> {
  const { data } = await api.post<Ticket>("/tickets", ticket);
  return data;
}

export async function updateTicketStatus(id: number, update: TicketUpdateStatus): Promise<Ticket> {
  const { data } = await api.patch<Ticket>(`/tickets/${id}/status`, update);
  return data;
}

export async function deleteTicket(id: number): Promise<void> {
  await api.delete(`/tickets/${id}`);
}

export async function loginAdmin(username: string, password: string): Promise<void> {
  const credentials = btoa(`${username}:${password}`);
  await api.post("/auth/login", null, {
    headers: { Authorization: `Basic ${credentials}` },
  });
}

File: backend/app/routers/tickets.py
from fastapi import APIRouter, Depends, Query

from ..auth import require_admin
from ..crud import create_ticket, delete_ticket, get_ticket, get_tickets, ticket_to_read
from ..database import get_db
from ..schemas import TicketCreate, TicketList, TicketRead, TicketUpdateStatus

router = APIRouter()


@router.post("", response_model=TicketRead, status_code=201)
def create(data: TicketCreate, db=Depends(get_db)):
    ticket = create_ticket(db, data)
    return ticket_to_read(ticket)


@router.get("", response_model=TicketList)
def list_tickets(
    status: str | None = Query(None),
    priority: str | None = Query(None),
    search: str | None = Query(None),
    sort_by: str = Query("created_at"),
    order: str = Query("desc", pattern="^(asc|desc)$"),
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    db=Depends(get_db),
):
    items, total = get_tickets(
        db,
        status=status,
        priority=priority,
        search=search,
        sort_by=sort_by,
        order=order,
        page=page,
        limit=limit,
    )
    return TicketList.build(
        items=[ticket_to_read(t) for t in items],
        total=total,
        page=page,
        limit=limit,
    )


@router.patch("/{ticket_id}/status", response_model=TicketRead)
def update_status(
    ticket_id: int,
    data: TicketUpdateStatus,
    db=Depends(get_db),
):
    from ..crud import update_ticket_status

    ticket = get_ticket(db, ticket_id)
    updated = update_ticket_status(db, ticket, data.status)
    return ticket_to_read(updated)


@router.delete("/{ticket_id}", status_code=204)
def delete(
    ticket_id: int,
    db=Depends(get_db),
    _admin: str = Depends(require_admin),
):
    ticket = get_ticket(db, ticket_id)
    delete_ticket(db, ticket)
